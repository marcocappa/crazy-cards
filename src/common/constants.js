export const CARD_TYPE = {
  STUDENT: 'student',
  ANYWHERE: 'anywhere',
  LIQUID: 'liquid',
};
