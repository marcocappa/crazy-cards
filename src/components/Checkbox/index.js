import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const Checkbox = ({
  className,
  name,
  value,
  isDisabled,
  label,
  isChecked,
  onChange,
}) => {
  const [checked, setChecked] = useState(isChecked);

  return (
    <>
      <input
        onChange={() => {
          onChange(value);
        }}
        type="checkbox"
        defaultChecked={checked}
        name={`${name}[]`}
        className={`crazy-card-checkbox ${classnames(className)}`}
        id={value}
        disabled={isDisabled}
      />
      <label
        htmlFor={value}
        onClick={() => (checked ? setChecked(false) : setChecked(true))}
      >
        {label}
      </label>
    </>
  );
};

Checkbox.propTypes = {
  className: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
  isChecked: PropTypes.bool,
};

Checkbox.defaultProps = {
  className: undefined,
  isDisabled: false,
  isChecked: false,
};

export default Checkbox;
