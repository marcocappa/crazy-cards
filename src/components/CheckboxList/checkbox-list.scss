.crazy-card-checkbox-list {
    margin: 0;
    padding: 0;
    list-style: none;

    &__item {
        margin-bottom: 1rem;
    }

    &__checkbox {
        margin-right: 0.5rem;
        cursor: pointer;
    }

    & label {
        cursor: pointer;
        display: inline-block;
    }
}
