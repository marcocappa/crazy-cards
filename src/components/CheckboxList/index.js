import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from '../Checkbox';
import './checkbox-list.scss';

const CheckboxList = ({ name, items, onSelect }) => {
  if (items.length === 0) return null;
  return (
    <ul className="crazy-card-checkbox-list">
      {items.map((item) => (
        <li
          key={`list-item-${item.value}`}
          className="crazy-card-checkbox-list__item"
          aria-label="List item"
        >
          <Checkbox
            onChange={onSelect}
            className="crazy-card-checkbox-list__checkbox"
            name={name}
            value={item.value}
            isDisabled={item.isDisabled}
            label={item.label}
            isChecked={item.checked || false}
          />
        </li>
      ))}
    </ul>
  );
};

CheckboxList.propTypes = {
  name: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      isDisabled: PropTypes.bool,
      label: PropTypes.string,
      checked: PropTypes.bool,
    })
  ),
  onSelect: PropTypes.func.isRequired,
};

CheckboxList.defaultProps = {
  items: [],
};

export default CheckboxList;
