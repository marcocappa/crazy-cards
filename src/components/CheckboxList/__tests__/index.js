import React from 'react';
import { shallow } from 'enzyme';
import CheckboxList from '../index';
import Checkbox from '../../Checkbox';
import { CARD_TYPE } from '../../../common/constants';

describe('<CheckboxList />', () => {
  const onSelect = jest.fn();
  const props = {
    name: 'list',
    items: [
      {
        value: CARD_TYPE.STUDENT,
        isDisabled: false,
        label: 'Student Card',
        checked: false,
      },
      {
        value: CARD_TYPE.LIQUID,
        isDisabled: false,
        label: 'Liquid Card',
        checked: false,
      },
    ],
    onSelect,
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CheckboxList {...props} />);
  });

  it('should match snapshot without items', () => {
    wrapper = shallow(
      <CheckboxList name="list-without-items" onSelect={onSelect} items={[]} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ul').length).toEqual(0);
  });

  it('should match snapshot with items', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ul').children().length).toEqual(props.items.length);
  });

  it('should invoke onSelect', () => {
    wrapper
      .find('.crazy-card-checkbox-list__checkbox')
      .at(0)
      .simulate('change');
    expect(onSelect).toHaveBeenCalledTimes(1);

    wrapper
      .find('.crazy-card-checkbox-list__checkbox')
      .at(1)
      .simulate('change');
    expect(onSelect).toHaveBeenCalledTimes(2);
  });
});
