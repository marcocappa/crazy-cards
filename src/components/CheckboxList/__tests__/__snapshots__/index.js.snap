// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxList /> should match snapshot with items 1`] = `
<ul
  className="crazy-card-checkbox-list"
>
  <li
    aria-label="List item"
    className="crazy-card-checkbox-list__item"
    key="list-item-student"
  >
    <Checkbox
      className="crazy-card-checkbox-list__checkbox"
      isChecked={false}
      isDisabled={false}
      label="Student Card"
      name="list"
      onChange={[MockFunction]}
      value="student"
    />
  </li>
  <li
    aria-label="List item"
    className="crazy-card-checkbox-list__item"
    key="list-item-liquid"
  >
    <Checkbox
      className="crazy-card-checkbox-list__checkbox"
      isChecked={false}
      isDisabled={false}
      label="Liquid Card"
      name="list"
      onChange={[MockFunction]}
      value="liquid"
    />
  </li>
</ul>
`;

exports[`<CheckboxList /> should match snapshot without items 1`] = `""`;
